libres/tests/enkf/enkf_obs_paths_detailed.cpp

```diff
@@ -0,0 +1,194 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+#include <string>
+#include <iostream>
+
+#include <catch2/catch.hpp>
+
+#include <ert/util/test_work_area.hpp>
+#include <ert/util/test_util.h>
+
+#include <ert/enkf/enkf_obs.hpp>
+
+/*
+ * Set up workarea, create and write start of file.
+ * Returns pointer to workares since it needs to be kept alive
+ * for the actual test
+ */
+ecl::util::TestArea *setup() {
+    ecl::util::TestArea *ta = new ecl::util::TestArea("conf");
+    util_make_path("obs_path");
+    FILE *stream = fopen("obs_path/conf.txt", "w");
+    fprintf(stream, "GENERAL_OBSERVATION WPR_DIFF_1 {"
+                    "DATA       = SNAKE_OIL_WPR_DIFF;"
+                    "INDEX_LIST = 400,800,1200,1800;"
+                    "RESTART    = 199;");
+    fclose(stream);
+    return ta; // We need to keep workarea alive over the test
+}
+
+/*
+ * Append keywords and write closing-bracket.
+ * Returns pointer to the parsed file
+ */
+conf_instance_type *write_conf(ecl::util::TestArea *workarea,
+                               std::vector<std::string> keyord_lines) {
+    FILE *stream = fopen("obs_path/conf.txt", "a");
+    for (auto s : keyord_lines)
+        fprintf(stream, "%s", s.data());
+    fprintf(stream, "};");
+    fclose(stream);
+
+    conf_class_type *enkf_conf_class = enkf_obs_get_obs_conf_class();
+    return conf_instance_alloc_from_file(enkf_conf_class, "enkf_conf",
+                                         "obs_path/conf.txt");
+}
+
+/*
+ * Just a convenient wrapper from char*  to std::string
+ */
+std::string get_path_error(conf_instance_type *enkf_conf) {
+    char *errors = conf_instance_get_path_error(enkf_conf);
+    if (errors) {
+        std::string s(errors);
+        free(errors);
+        return s;
+    } else
+        return std::string("");
+}
+
+/*
+ * Syntactic sugar for checking if a string contains another string
+ * The memmber-function ::contains() is introduced in C++23
+ * */
+bool contains(std::string errors, std::string kw) {
+    return errors.find(kw) != std::string::npos;
+}
+
+void touch_file(std::string name) {
+    FILE *stream = fopen(name.data(), "w");
+    fclose(stream);
+}
+
+TEST_CASE("Test parsing keywords in configuration", "[unittest]") {
+    GIVEN("Set up infrastructure") {
+        ecl::util::TestArea *workarea = setup();
+
+        THEN("No relevant keywords, no errors") {
+            conf_instance_type *enkf_conf = write_conf(workarea, {});
+            std::string errors = get_path_error(enkf_conf);
+            REQUIRE(!contains(errors, "OBS_FILE"));
+            REQUIRE(!contains(errors, "ERROR_COVAR"));
+            REQUIRE(!contains(errors, "INDEX_FILE"));
+            conf_instance_free(enkf_conf);
+        }
+
+        THEN("Single keyword, file NOT present") {
```

Comment:
If you like, the `GIVEN` and `THEN` keywords relate to BDD style which normally follow 3xA pattern (Assemble, Assert, Act) which map to the macros `GIVEN`, `WHEN` and `THEN` which are all more or less an alias of `SECTION`.

To use such a 3xA style I would suggest:
```cpp
TEST_CASE("Test parsing keywords in configuration", "[unittest]") {
    GIVEN("A workearea") {
        ecl::util::TestArea *workarea = setup();
        GIVEN("A conf file with no relevant keywords") {
            conf_instance_type *enkf_conf = write_conf(workarea, {});
            THEN("There are no errors") {
                std::string errors = get_path_error(enkf_conf);
                REQUIRE(!contains(errors, "OBS_FILE"));
                REQUIRE(!contains(errors, "ERROR_COVAR"));
                REQUIRE(!contains(errors, "INDEX_FILE"));
            }
            conf_instance_free(enkf_conf);
        }

        GIVEN("A conf file with OBS_FILE"){
            conf_instance_type *enkf_conf =
                write_conf(workarea, {"OBS_FILE   = obs.txt;"});
            WHEN("There is no OBS_FILE present"){
                THEN("Error refers to missing file") {
                    std::string errors = get_path_error(enkf_conf);
                    REQUIRE(contains(errors, "OBS_FILE=>" + workarea->test_cwd() +
                                                 "/obs_path/obs.txt"));
                    REQUIRE(!contains(errors, "ERROR_COVAR"));
                    REQUIRE(!contains(errors, "INDEX_FILE"));
                }
            }
            WHEN("There is an OBS_FILE present") {
                touch_file("obs_path/obs.txt");
                THEN("There are no errors") {
                    std::string errors = get_path_error(enkf_conf);
                    REQUIRE(!contains(errors, "OBS_FILE"));
                    REQUIRE(!contains(errors, "ERROR_COVAR"));
                    REQUIRE(!contains(errors, "INDEX_FILE"));
                }
            }
            conf_instance_free(enkf_conf);
        }


        GIVEN("A conf file with OBS_FILE in subdir obs_path") {
            conf_instance_type *enkf_conf =
                write_conf(workarea, {"OBS_FILE   = obs_path/obs.txt;"});

            WHEN("The OBS_FILE is in the wrong location"){
                touch_file("obs_path/obs.txt");
                THEN("Error refers to missing file"){
                    std::string errors = get_path_error(enkf_conf);
                    REQUIRE(contains(errors, "OBS_FILE=>" + workarea->test_cwd() +
                                                 "/obs_path/obs_path/obs.txt"));
                    REQUIRE(!contains(errors, "ERROR_COVAR"));
                    REQUIRE(!contains(errors, "INDEX_FILE"));
                }
            }
            WHEN("The OBS_FILE is in the subdirectory"){
                util_make_path("obs_path/obs_path");
                touch_file("obs_path/obs_path/obs.txt");
                THEN("There are no errors") {
                    std::string errors = get_path_error(enkf_conf);
                    REQUIRE(!contains(errors, "OBS_FILE"));
                    REQUIRE(!contains(errors, "ERROR_COVAR"));
                    REQUIRE(!contains(errors, "INDEX_FILE"));
                }
            }
            conf_instance_free(enkf_conf);
        }

        GIVEN("A conf file with OBS_FILE, ERROR_COVAR and INDEX_FILE keywords") {
            conf_instance_type *enkf_conf =
                write_conf(workarea, {"OBS_FILE    = obs.txt;"
                                      "ERROR_COVAR = covar.txt;"
                                      "INDEX_FILE  = index.txt;"});

            WHEN("Only OBS_FILE exists"){
                touch_file("obs_path/obs.txt");
                THEN("Error refer to missing covar and index file") {
                    std::string errors = get_path_error(enkf_conf);
                    REQUIRE(!contains(errors, "OBS_FILE"));
                    REQUIRE(contains(errors, "ERROR_COVAR=>" + workarea->test_cwd() +
                                                 "/obs_path/covar.txt"));
                    REQUIRE(contains(errors, "INDEX_FILE=>" + workarea->test_cwd() +
                                                 "/obs_path/index.txt"));
                }
            }
            WHEN("covar, obs, and index files exist"){
                touch_file("obs_path/obs.txt");
                touch_file("obs_path/covar.txt");
                touch_file("obs_path/index.txt");
                THEN("There are no errors") {
                    std::string errors = get_path_error(enkf_conf);
                    REQUIRE(!contains(errors, "OBS_FILE"));
                    REQUIRE(!contains(errors, "ERROR_COVAR"));
                    REQUIRE(!contains(errors, "INDEX_FILE"));
                }
            }
            conf_instance_free(enkf_conf);
        }
        delete workarea;
    }
}

```

Reply:
Thanks for tip (and for the rewrite 😄  ) I|ll adopt that style.